### PR TITLE
KOGITO-5225/KOGITO-5154: Support React components on Kogito Java-editors

### PR DESCRIPTION
**Pull request by karreiro**
### JIRAs

- [KOGITO-5225](https://issues.redhat.com/browse/KOGITO-5225): New boxed expression editor - React Components module pipeline

- [KOGITO-5154](https://issues.redhat.com/browse/KOGITO-5154): Import Java classes as data objects on Kogito - Integrate with GWT

### Description

This PR introduces the infrastructure for supporting React components on Kogito editors (DMN and BPMN). They live in the `packages` directory and follow the convention of **components** and **loaders**:
- **Components** - all elements in the editors (a marshaller, a dialog, or even a FEEL parser)
- **Loaders** - enable editors to load and render their components

The included packages are:
- **`packages/hello-world-component`** - it's a simple component used for testing purposes (my following PR will replace it with the `boxed-expression-editor`)
- **`packages/dmn-loader`** - responsible for loading components in the DMN editor

### How to test this PR

You may render the **Hello world** component to test this PR locally. Add the following lines the `DMNClientEntryPoint.java`, and run the `kie-wb-common-dmn-webapp-kogito-testing` to see the React component rendering in the GWT world:

```java
final Element container = DomGlobal.document.createElement("div");

container.id = "container";
DomGlobal.document.body.appendChild(container);

DMNLoader.renderFeelDialog("#container");
```

Alternatively, you may run the `kie-wb-common-dmn-webapp-kogito-testing` and paste this on browser console:

```javascript
const container = document.createElement("div");
container.id = "container";
document.body.appendChild(container);
__KIE__DMN_LOADER__.renderHelloWorld("#container");
```

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* Retest PR: <b>jenkins retest this</b>
* A full downstream build: <b>jenkins do fdb</b>
* A compile downstream build: <b>jenkins do cdb</b>
* A full production downstream build: <b>jenkins do product fdb</b>
* An upstream build: <b>jenkins do upstream</b>
</details>

### CODE REVIEW

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-testing/pom.xml [hunk 1]
```diff
@@ -522,6 +528,27 @@
...
+            <executions>
+              <execution>
+                <id>add-source</id>
+                <phase>generate-resources</phase>
+                <goals>
+                  <goal>add-source</goal>
+                </goals>
+                <configuration>
+                  <sources>
+                    <source>/Volumes/Navi/Projects/forks/kogito-editors-java/kogito-editors-js</source>
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-testing/pom.xml:545]: Is this correct?

Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-testing/pom.xml:545]: Thanks, @manstis! I missed this directory here. I'm fixing in a separate commit. Thanks for checking it! :-)

Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-testing/pom.xml]: Actually, I've applied the fix in a squashed commit ;-) Thanks again for checking, Mike!

**Review by tiagobento — APPROVED** [review]
Approving this as it provides a good starting point for integrating React components into the Editors. Great job @karreiro!

**Review by vpellegrino — APPROVED** [review]
Congrats @karreiro 
Very nice job!
I also enjoyed the Node script `new-component.js` for a component generation :100:

**Review by yesamer — APPROVED** [review]
Excellent

**Review by manstis — APPROVED** [review]
LGTM, my comment related only to a hard coded path.

File: kogito-editors-js/pom.xml [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+      <!-- Build kogito-editors-js components -->
+      <plugin>
+        <groupId>com.github.eirslett</groupId>
+        <artifactId>frontend-maven-plugin</artifactId>
+        <configuration>
+          <installDirectory>${project.build.directory}</installDirectory>
+          <nodeVersion>v16.2.0</nodeVersion>
+          <npmVersion>7.15.1</npmVersion>
+          <yarnVersion>v1.22.10</yarnVersion>
```
Comment (domhanak) [inline kogito-editors-js/pom.xml:28]: Could we please set these up in parent pom.xml? Would be easier to update in the future.

**Review by domhanak — APPROVED** [review]
Very interesting!

File: kogito-editors-js/docs/new-component.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+> **Notice**: The directory already has the proper suffix by convention!
+
+Now you may run this to install the default dependencies of your component and development web app:
+
+```bash
+yarn install
+
+cd showcase
+yarn install
```
Comment (jomarko) [inline kogito-editors-js/docs/new-component.md:37]: maybe we should use `yarn bootstrap` as we use in `kogito-tooling`? or here is the `install` autogenerated and we can not change it?

File: kogito-editors-js/docs/new-component.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+Finally, you're ready to run the web app and see your component alive:
+
+```
+yarn start
+```
+
+### 3. Expose with the a Loader
+
+Now, we need to get this component in some loader. So, your GWT editor (DMN or BPMN) will be able to render it.
```
Comment (jomarko) [inline kogito-editors-js/docs/new-component.md:48]: scesim?

File: kogito-editors-js/docs/new-component.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+- And, finally, add the following snippet in the `src/index.tsx` file:
+
+```typescript
+import { FeelDialog } from "feel-dialog-component";
+
+const renderFeelDialog = (selector: string) => {
+  ReactDOM.render(<FeelDialog />, document.querySelector(selector));
+};
+
+export { renderFeelDialog };
```
Comment (jomarko) [inline kogito-editors-js/docs/new-component.md:63]: I am getting this error at this point
Screenshot from 2021-07-02 09-25-42

**Review by jomarko — COMMENTED** [review]
I was not able to finish instructions for new component, any advice?

**Comment by jomarko** [comment]
Could we please also apply code formatting https://github.com/karreiro/kogito-editors-java/pull/12

File: kogito-editors-js/README.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
+# Kogito editors JS
+
+**Kogito editors JS** module holds the next generation of JavaScript/TypeScript/React components for the **DMN** and **BPMN** editors.
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kogito-editors-js/README.md]: Fix applied :-) Thanks!

File: kogito-editors-js/docs/new-component.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
+# How to create a Kogito editor React component
+
+All Kogito editor React components live in the `packages` directory and follow the convention of **components** and **loaders**.
+
+- **Components** - all elements an editor may be a marshaller, a dialog, a Boxed Expression editor, or even a FEEL editor.
+
+- **Loaders** - this is more specific. It teaches how each editor (DMN or BPMN) may load their components.
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kogito-editors-js/docs/new-component.md]: Fix applied :-) Thanks!

File: kogito-editors-js/docs/new-component.md (hunk 4 above)
Comment (karreiro) [reply kogito-editors-js/docs/new-component.md]: Fix applied :-) Thanks!

File: kogito-editors-js/docs/new-component.md (hunk 5 above)
Comment (karreiro) [reply kogito-editors-js/docs/new-component.md]: Fix applied :-) Thanks!

File: kogito-editors-js/docs/new-component.md (hunk 3 above)
Comment (karreiro) [reply kogito-editors-js/docs/new-component.md:37]: We're at a sub-package level in this context, and we're relying on one of Yarn's default commands (`install`) to proceed.

But, we do have the `yarn bootstrap` script that relies on Lerna (just like Kogito Tooling) at the root level :-)

**Review by romartin — APPROVED** [review]

File: kogito-editors-js/pom.xml (hunk 2 above)
Comment (karreiro) [reply kogito-editors-js/pom.xml:28]: Sure! Thanks for the suggestion @domhanak. I've created this [Jira](https://issues.redhat.com/browse/KOGITO-5466) to not lose track of this enhancement.

**Comment by danielzhe** [comment]
I think this PR is causing an error with `mvn clean install`:

[INFO] --- frontend-maven-plugin:1.8.0:yarn (yarn build:prod) @ kogito-editors-js ---
[INFO] Running 'yarn build:prod' in F:\Dev\kogito-editors-java\kogito-editors-js
[INFO] yarn run v1.22.10
[INFO] $ rimraf dist && yarn format:check && lerna run build:prod --stream
[INFO] $ prettier --check .
[INFO] Checking formatting...
[INFO] [warn] .prettierrc.json
[INFO] [warn] docs\new-component.md
[INFO] [warn] docs\wire.md
[INFO] [warn] lerna.json
[INFO] [warn] package.json
[INFO] [warn] packages\.template-component\package.json
[INFO] [warn] packages\.template-component\showcase\config-overrides.js
[INFO] [warn] packages\.template-component\showcase\package.json
[INFO] [warn] packages\.template-component\showcase\public\index.html
[INFO] [warn] packages\.template-component\showcase\public\manifest.json
[INFO] [warn] packages\.template-component\showcase\src\index.css
[INFO] [warn] packages\.template-component\showcase\src\index.tsx
[INFO] [warn] packages\.template-component\showcase\src\react-app-env.d.ts
[INFO] [warn] packages\.template-component\showcase\tsconfig.json
[INFO] [warn] packages\.template-component\src\index.css
[INFO] [warn] packages\.template-component\src\index.tsx
[INFO] [warn] packages\.template-component\src\lib\ItWorks.css
[INFO] [warn] packages\.template-component\src\lib\ItWorks.tsx
[INFO] [warn] packages\.template-component\tsconfig.json
[INFO] [warn] packages\.template-component\webpack.development.js
[INFO] [warn] packages\.template-component\webpack.production.js
[INFO] [warn] packages\dmn-loader\package.json
[INFO] [warn] packages\dmn-loader\src\index.tsx
[INFO] [warn] packages\dmn-loader\src\wire.ts
[INFO] [warn] packages\dmn-loader\tsconfig.json
[INFO] [warn] packages\dmn-loader\webpack.development.js
[INFO] [warn] packages\dmn-loader\webpack.production.js
[INFO] [warn] packages\hello-world-component\package.json
[INFO] [warn] packages\hello-world-component\showcase\config-overrides.js
[INFO] [warn] packages\hello-world-component\showcase\package.json
[INFO] [warn] packages\hello-world-component\showcase\public\index.html
[INFO] [warn] packages\hello-world-component\showcase\public\manifest.json
[INFO] [warn] packages\hello-world-component\showcase\src\index.css
[INFO] [warn] packages\hello-world-component\showcase\src\index.tsx
[INFO] [warn] packages\hello-world-component\showcase\src\react-app-env.d.ts
[INFO] [warn] packages\hello-world-component\showcase\tsconfig.json
[INFO] [warn] packages\hello-world-component\src\index.css
[INFO] [warn] packages\hello-world-component\src\index.tsx
[INFO] [warn] packages\hello-world-component\src\lib\HelloWorld.css
[INFO] [warn] packages\hello-world-component\src\lib\HelloWorld.tsx
[INFO] [warn] packages\hello-world-component\tsconfig.json
[INFO] [warn] packages\hello-world-component\webpack.development.js
[INFO] [warn] packages\hello-world-component\webpack.production.js
[INFO] [warn] README.md
[INFO] [warn] scripts\new-component.js
[INFO] [warn] target\node\yarn\dist\bin\yarn.js
[INFO] [warn] target\node\yarn\dist\lib\cli.js
[INFO] [warn] target\node\yarn\dist\lib\v8-compile-cache.js
[INFO] [warn] target\node\yarn\dist\preinstall.js
[INFO] [warn] target\node\yarn\dist\README.md
[INFO] [warn] tsconfig.json
[INFO] [warn] webpack.common.development.js
[INFO] [warn] webpack.common.js
[INFO] [warn] webpack.common.production.js
[INFO] [warn] Code style issues found in the above file(s). Forgot to run Prettier?
[INFO] error Command failed with exit code 1.
[INFO] info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
[INFO] error Command failed with exit code 1.
[INFO] info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

Can you please confirm? @karreiro

**Comment by jomarko** [comment]
@danielzhe is this log coming from fresh master or from some work in progress you rebased with master? if it is the second one, please run `yarn run format` in `kogito-editors-js` before you run `mvn clean install` in `kogito-editors-java`

**Comment by karreiro** [comment]
@danielzhe Actually the PR passes locally and in the CI, see:



I'm pinging you on GChat to get more details about your environment, then we can make it run (and update some README with the steps to solve the issue, if necessary) :-)

**Comment by danielzhe** [comment]
Thank you. It's fresh master, @jomarko . I'm checking with @karreiro in Gchat to see if it is something in my environment.

**Pull request merged**